File: secondExercise_2/src/main/java/DiscountPolicy.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

abstract public class DiscountPolicy {
    private List<DiscountCondition> conditions = new ArrayList<>();

    public DiscountPolicy(DiscountCondition ... conditions) {
        this.conditions = Arrays.asList(conditions);
    }

    public final double calculateDiscountFee(Screen screen) {
        for(DiscountCondition each: conditions) {
            if (each.isSatisfied(screen)) {
                return this.getDiscountFee(screen);
            }
        }
        return 0;
    }

    protected abstract double getDiscountFee(Screen screen);

}

File: secondExercise/src/main/kotlin/DiscountPolicy.kt
abstract class DiscountPolicy(private val discountConditions: List<DiscountCondition>) {
    fun calculateDiscountFee(screen: Screen): Double {
        for(discountCondition in discountConditions) {
            if(discountCondition.isSatisfied(screen)) return getDiscountFee(screen.fee)
        }
        return 0.0
    }

    abstract fun getDiscountFee(fee: Double): Double
}
